File: 2024/Go/solutions/day14.go
package solutions

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

func getPositionAftertSeconds(pos []int,vel []int,maxX,maxY int,t int) []int{
    newX := pos[0] + (t * vel[0]) + (t * maxX)
    newY := pos[1] + (t * vel[1]) + (t * maxY)
    newX %= maxX
    newY %= maxY
    return []int{newX,newY}
}

func getSafetyFactor(positions,velocities [][]int,maxX,maxY int,t int) (int,[][]int){
    q1,q2,q3,q4 := 0,0,0,0
    //       |
    //  Q1   | Q2
    // ------------
    //  Q3   | Q4
    //       |
    newPositions := [][]int{}
    for i := range(len(positions)){
        newPos := getPositionAftertSeconds(positions[i],velocities[i],maxX,maxY,t)
        newPositions = append(newPositions,newPos)
        if newPos[0] < (maxX/2) && newPos[1] < (maxY/2){
            q1 += 1
        }
        if newPos[0] > (maxX/2) && newPos[1] < (maxY/2){
            q2 += 1
        }
        if newPos[0] < (maxX/2) && newPos[1] > (maxY/2){
            q3 += 1
        }
        if newPos[0] > (maxX/2) && newPos[1] > (maxY/2){
            q4 += 1
        }
    }
    return q1*q2*q3*q4,newPositions
}
func printGrid(positions [][]int,maxX,maxY int){
    grid := make([][]int,maxY+1,maxY+1)
    for y := range(maxY){
        grid[y] = make([]int,maxX+1,maxX+1)
    }
    for _,position := range(positions){
        grid[position[1]][position[0]] = 1
    }
    for y := range(maxY){
        for x := range(maxX){
            if grid[y][x] == 0{
                fmt.Print(" ")
            }else{
                fmt.Print("*")
            }
        }
        fmt.Println()
    }
}

func Day14(){
    file,err := os.Open("../input.txt")
    if err != nil{
        log.Fatalf("Error while opening the file: %v",err)
    }
    defer file.Close()
    scanner := bufio.NewScanner(file)
    positions := [][]int{}
    velocities := [][]int{}
    for (scanner.Scan()){
        splitBySpace := strings.Split(scanner.Text()," ")
        pos := strings.Split(splitBySpace[0],",")
        posX,_ := strconv.Atoi(strings.Trim(pos[0],"p="))
        posY,_ := strconv.Atoi(pos[1])
        positions = append(positions, []int{posX,posY})
        vel := strings.Split(splitBySpace[1],",")
        velX,_ := strconv.Atoi(strings.Trim(vel[0],"v="))
        velY,_ := strconv.Atoi(vel[1])
        velocities = append(velocities, []int{velX,velY})
    } 
    maxX := 101
    maxY := 103
    part1,_ := getSafetyFactor(positions,velocities,maxX,maxY,100)
    fmt.Printf("Part 1:%v\n",part1)
    for i := range(500){
        safetyFactor,newPositions := getSafetyFactor(positions,velocities,maxX,maxY,(7500 + i))
        // fmt.Println(safetyFactor)
        if safetyFactor <= 110 * 110 * 100 * 100{
            fmt.Printf("------------------------------------------------------- Time %v -----------------------------------------------------------------------------}\n",i)
            printGrid(newPositions,maxX,maxY)
        }
    }
}

File: 2024/Zig/src/day14.zig
const std = @import("std");

const maxX: i32 = 101;
const maxY: i32 = 103;

fn getNewPosition(position: [2]i32, velocity: [2]i32, t: i32) [2]i32 {
    return [2]i32{ @mod((position[0] + velocity[0] * t + maxX * t), maxX), @mod((position[1] + velocity[1] * t + maxY * t), maxY) };
}

fn getSafetyScore(positions: std.ArrayList([2]i32), velocities: std.ArrayList([2]i32), t: i32, newPositions: *std.ArrayList([2]i32)) !u32 {
    newPositions.clearRetainingCapacity();
    const length = positions.items.len;
    var q1: u32 = 0;
    var q2: u32 = 0;
    var q3: u32 = 0;
    var q4: u32 = 0;
    for (0..length) |idx| {
        const newPos = getNewPosition(positions.items[idx], velocities.items[idx], t);
        try newPositions.append(newPos);
        if (newPos[0] < maxX / 2 and newPos[1] < maxY / 2) {
            q1 += 1;
        }
        if (newPos[0] > maxX / 2 and newPos[1] < maxY / 2) {
            q2 += 1;
        }
        if (newPos[0] < maxX / 2 and newPos[1] > maxY / 2) {
            q3 += 1;
        }
        if (newPos[0] > maxX / 2 and newPos[1] > maxY / 2) {
            q4 += 1;
        }
    }
    return q1 * q2 * q3 * q4;
}

fn printGrid(allocator: std.mem.Allocator, positions: std.ArrayList([2]i32)) !void {
    var grid = try allocator.alloc([]u8, maxY + 1);
    for (0..maxY) |y| {
        grid[y] = try allocator.alloc(u8, maxX + 1);
    }
    defer {
        for (0..maxY) |y| {
            allocator.free(grid[y]);
        }
        allocator.free(grid);
    }
    for (0..maxY) |y| {
        for (0..maxX) |x| {
            grid[y][x] = ' ';
        }
    }
    for (positions.items) |pos| {
        grid[@intCast(pos[1])][@intCast(pos[0])] = '*';
    }
    for (0..maxY) |y| {
        std.debug.print("{s}\n", .{grid[y]});
    }
}

fn day14_sol(allocator: std.mem.Allocator) !void {
    const file = try std.fs.cwd().openFile("../input.txt", .{});
    defer file.close();
    var positions = std.ArrayList([2]i32).init(allocator);
    defer positions.deinit();
    var velocities = std.ArrayList([2]i32).init(allocator);
    defer velocities.deinit();
    while (try file.reader().readUntilDelimiterOrEofAlloc(allocator, '\n', std.math.maxInt(u64))) |line| {
        defer allocator.free(line);
        var spaceIt = std.mem.splitScalar(u8, line, ' ');
        var idx: u32 = 0;
        while (spaceIt.next()) |token| {
            if (idx == 0) {
                var pos = std.mem.splitScalar(u8, std.mem.trimLeft(u8, token, "p="), ',');
                try positions.append([2]i32{ try std.fmt.parseInt(i32, pos.next().?, 10), try std.fmt.parseInt(i32, pos.next().?, 10) });
                idx += 1;
            } else {
                var vel = std.mem.splitScalar(u8, std.mem.trim(u8, token, "v="), ',');
                try velocities.append([2]i32{ try std.fmt.parseInt(i32, vel.next().?, 10), try std.fmt.parseInt(i32, vel.next().?, 10) });
            }
        }
    }
    var newPostions = std.ArrayList([2]i32).init(allocator);
    defer newPostions.deinit();
    const part1: u32 = try getSafetyScore(positions, velocities, 100, &newPostions);
    std.debug.print("Part 1: {d}\n", .{part1});
    //Already implemented in Go, so I know the range here. The safety score is low for the case where we have an XMAS tree as the points be close together. Total 500 points
    //Trail and error on the safety score
    for (7500..8000) |i| {
        const safetyScore = try getSafetyScore(positions, velocities, @intCast(i), &newPostions);
        if (safetyScore <= 110 * 110 * 100 * 100) {
            std.debug.print("--------------------------------------------------Time = {d}--------------------------------------------------\n", .{i});
            try printGrid(allocator, newPostions);
        }
    }
}

test "day14" {
    try day14_sol(std.testing.allocator);
}
